plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+       , AsFutureError e
+       )
+    => Future
+    -> Contract s e FutureOwners
+initialiseFuture future = do
+    (s, ownRole) <- endpoint @"initialise-future" @(FutureSetup, Role)
+    ftos <- setupTokens
+    let escr = escrowParams future ftos s
+        escrowPayment = Escrow.payRedeemRefund escr (initialMargin future <> tokenFor Long ftos <> tokenFor Short ftos)
+    e <- withContractError (review _EscrowFailed) escrowPayment
+    either (throwing _EscrowRefunded) (\_ -> pure ftos) e
+
+settleFuture
+    :: ( HasEndpoint "settle-future" (OracleValue Value) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+settleFuture future@Future{ftDeliveryDate} ftos = do
+    ov <- endpoint @"settle-future"
+    let address = futureAddress future ftos
+    unspentOutputs <- utxoAt address
+    accounts <- currentBalances future ftos
+    let tx = payoutsTx (payouts future accounts (ovValue ov)) ftos
+             Haskell.<> Tx.collectFromScript unspentOutputs (validatorScript future ftos) (Ledger.RedeemerScript $ PlutusTx.toData $ Settle ov)
+                & validityRange .~ Interval.from (succ ftDeliveryDate)
+    void $ writeTx tx
+
+settleEarly
+    :: ( HasEndpoint "settle-early" (OracleValue Value) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+settleEarly future@Future{ftDeliveryDate} ftos = do
+    ov <- endpoint @"settle-early"
+    let address = futureAddress future ftos
+        spotPrice = ovValue ov
+        minMargin = requiredMargin future spotPrice
+    unspentOutputs <- utxoAt address
+    MarginAccounts{ftsShortMargin, ftsLongMargin} <- currentBalances future ftos
+    
+    violatingRole <- 
+        if ftsShortMargin `lt` minMargin then pure Short
+        else if ftsLongMargin `lt` minMargin then pure Long
+        else throwing_ _MarginRequirementsNotViolated
+    
+    let payment = case violatingRole of
+                    Long -> Payouts{payoutsLong=mempty, payoutsShort = ftsShortMargin + ftsLongMargin }
+                    Short -> Payouts{payoutsLong=ftsShortMargin+ftsLongMargin, payoutsShort = mempty }
+        tx = payoutsTx payment ftos
+             Haskell.<> Typed.collectFromScript unspentOutputs (scriptInstance future ftos) (SettleEarly ov)
+    void $ writeTx tx
+
+-- | Determine the current 'MarginAccounts' by looking at the unspent output
+--   of the contract.
+currentBalances
+    :: ( HasUtxoAt s
+       , AsFutureError e)
+    => Future
+    -> FutureOwners
+    -> Contract s e MarginAccounts
+currentBalances future ftos = do
+    let address = futureAddress future ftos
+    unspentOutputs <- utxoAt address
+    theOutput <- case toList (AM.outRefMap unspentOutputs) of
+                    []  -> throwing_ _FutureUtxoNotFound
+                    [x] -> pure x
+                    _   -> throwing_ _FutureUtxoNotUnique
```

Comment:
So this is vulnerable to contract poisoning. Worth a note at least.

Reply:
Good point. I think this can be fixed by (1) checking that the value on the output is equal to the margin account (in this case poisoning the contract doesn't cause any financial damage because the attacker wouldn't gain anything), or (2) adding a token that's always attached to the current output

Reply:
Well, it could still be in an unexpected state. I think (2) is the way to go in the long run, we'll want it for state machines too, I imagine.